### PR TITLE
Sudo Plugin Model

### PR DESCRIPTION
**Pull request by amitlimaye**
This PR is part of a project that bring the following changes:

- [x] A new policy named `UserAccessPolicy` allows to define who can sudo where. 
_Example: `user=amit` (subject) is allowed to do `sudo mysql` and `sudo docker` (sudo users) in host where `$enforcerid=X` (object)_

- [x] Rendered new ingress policy rules :

Example:
```
    "ingressPolicies": {
      "filePolicyRules": null,
      "isolationPolicyRules": null,
      "networkPolicyRules": [
        ...
      ],
      "userAccessRules": [
        {
          "ID": "5d0045296eb576bfc0b70426",
          "action": {
            "action=user-access-info": {
              "allowedSudoUsers": [
                "root"
              ]
            }
          },
          "name": "chris-test",
          "policyNamespace": "/",
          "policyUpdateTime": "2019-06-12T00:20:01.769Z",
          "propagated": false,
          "relation": [],
          "tagClauses": [
            [
              "$identity=enforcer"
            ]
          ]
        }
      ]
    },
```

- [x] Added a policy renderer for UserAccess policies:

```
% apoctl api create policyrenderer -k type UserAccess -k tags '["$identity=enforcer"]' -o json -k processMode Object
{
  "policies": [
    {
      "ID": "5d014e3b6eb5762e67f76899",
      "action": {
        "action=user-access-info": {
          "allowedSudoUsers": [
            "root"
          ]
        }
      },
      "name": "chris-test",
      "policyNamespace": "/apomux",
      "policyUpdateTime": "2019-06-12T19:10:51.492Z",
      "propagated": false,
      "relation": [],
      "tagClauses": [
        [
          "test=chris"
        ]
      ]
    }
  ],
  "processMode": "Object",
  "tags": [
    "$identity=enforcer"
  ],
  "type": "UserAccess"
}
```

Part of project https://github.com/orgs/aporeto-inc/projects/475

### CODE REVIEW

File: specs/privilegeescalationpolicy.spec [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    global_parameters:
+    - $filtering
+  extends:
+  - '@base'
+  - '@described'
+  - '@disabled'
+  - '@identifiable-not-stored'
+  - '@metadatable'
+  - '@named'
+  - '@propagated'
```
Comment (primalmotion) [inline specs/privilegeescalationpolicy.spec]: is it propagating?

File: specs/useraccesspolicy.spec [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  resource_name: useraccesspolicies
+  entity_name: UserAccessPolicy
+  package: squall
+  group: policy/access
+  description: The enforcer policy controls user access.
+  aliases:
+  - usrpol
+  - usrpols
+  get:
+    description: Retrives the UserAccessPolicy with the given ID.
```
Comment (primalmotion) [inline specs/useraccesspolicy.spec]: typo

File: specs/useraccesspolicy.spec [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  - '@timeable'
+
+# Indexes
+indexes:
+- - :no-inherit
+
+# Attributes
+attributes:
+  v1:
+  - name: allowSudoUsers
```
Comment (primalmotion) [inline specs/useraccesspolicy.spec]: `allowedSudoUsers` we never use verb in properties

File: specs/useraccesspolicy.spec (hunk 2 above)
Comment (t00f) [reply specs/useraccesspolicy.spec]: Fixed

File: specs/useraccesspolicy.spec (hunk 3 above)
Comment (t00f) [reply specs/useraccesspolicy.spec]: Done

File: specs/privilegeescalationpolicy.spec (hunk 1 above)
Comment (t00f) [reply specs/privilegeescalationpolicy.spec]: I think it needs to propagate down to child namespaces yes...

**Comment by t00f** [comment]
I also added the default_value of `AllowedSudoUsers` to be []string{"root"}.

**Comment by primalmotion** [comment]
If you give that default value, you will not be able to reset to nothing. Plus maximum power by default is not vey aporeto-esque

**Comment by t00f** [comment]
Ok so you would prefer not having it. I will set it as an example value then

**Comment by t00f** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "d524f145ddd575fc839fb9e7fe2a16ada4f46bdc"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "84e5c13fd2721054d93971e03931c0ceab3fb4bb"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "bf8b1717c6280c9f3a0f50aad497bfcea68b1d80"
  }
]
```

**Comment by t00f** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "84e5c13fd2721054d93971e03931c0ceab3fb4bb"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "bf8b1717c6280c9f3a0f50aad497bfcea68b1d80"
  },
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "545b46717f918a80d2da6fb428bdb1bde1e9482e"
  }
]
```

**Comment by t00f** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "b01f66611a23d5ece12b202b2cf2f3b2d814031e"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "8b63105319067cf7930f25c7d193a5bdb5795bc6"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "f499ac72c11bf3824d8df884cbb235738037c24d"
  }
]
```

**Comment by t00f** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "8b63105319067cf7930f25c7d193a5bdb5795bc6"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "f499ac72c11bf3824d8df884cbb235738037c24d"
  },
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "2be8e4fbd68acd4fc6afab1b46dbfe30b42c64ba"
  }
]
```

**Comment by t00f** [comment]
_Quick note to on default value_

Integration test of backend was failing because of :
```
Iteration [1] log after 0s
  create a useraccesspolicy
  - [PASS] connecting to events channel for 'create' event for 'useraccesspolicy' should work
  - [PASS] associatedTags should contain tag=a
  - [PASS] normalizedTags should contain tag=a
  - [PASS] normalizedTags should contain the $identity tag
  - [PASS] normalizedTags should contain the $namespace tag
  - [PASS] namespace should be correct
  - [PASS] annotations should be correct
  - [PASS] pushed object and returned object should be identical
  - [PASS] name is correct
  - [PASS] description is correct
  - [PASS] object is correct
  - [PASS] subject is correct
  - [PASS] propagate is correct
  
  error: [FAIL] allowedSudoUsers is correct: expected: '[root]', actual '[root]'
```

**Comment by dstiliadis** [comment]
An empty list means everyone for the sudousers .. That would be my interpretation to increase usability rather than defaults.

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "1d076cdfe68ef73716393de063f9e337e2a683b1"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "8b63105319067cf7930f25c7d193a5bdb5795bc6"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "f499ac72c11bf3824d8df884cbb235738037c24d"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "c15b07cee3a156e79334afeb17ff33817486b72c"
  }
]
```

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "8b63105319067cf7930f25c7d193a5bdb5795bc6"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "f499ac72c11bf3824d8df884cbb235738037c24d"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "4ae1f2f1250d717c9420af9c5b3577e0c8d509ad"
  },
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "1d076cdfe68ef73716393de063f9e337e2a683b1"
  }
]
```

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "d9312c8498f76a3b013efcdbaf7bf7aa30117019"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "e42553544c19ecb0ccb5d012902e2871a7aef38f"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "1fc4f1a1820fd44a3f0dc6bb97851517cf537187"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "4ae1f2f1250d717c9420af9c5b3577e0c8d509ad"
  },
  {
    "project": "sudopolicy",
    "component": "domingo",
    "pr-id": "70",
    "commit-sha": "bfa0d53cbe0e9186a17b6919e180d593e047f70f"
  }
]
```

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "1fc4f1a1820fd44a3f0dc6bb97851517cf537187"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "c81c8f08b561e9020eb5a289999dba8f04a71472"
  },
  {
    "project": "sudopolicy",
    "component": "domingo",
    "pr-id": "70",
    "commit-sha": "bfa0d53cbe0e9186a17b6919e180d593e047f70f"
  },
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "d9312c8498f76a3b013efcdbaf7bf7aa30117019"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "e42553544c19ecb0ccb5d012902e2871a7aef38f"
  }
]
```

**Comment by t00f** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "1fc4f1a1820fd44a3f0dc6bb97851517cf537187"
  },
  {
    "project": "sudopolicy",
    "component": "clad",
    "pr-id": "1163",
    "commit-sha": "7ca731dddc48c18f9025ccf0606639e7ba5c7681"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "c81c8f08b561e9020eb5a289999dba8f04a71472"
  },
  {
    "project": "sudopolicy",
    "component": "domingo",
    "pr-id": "70",
    "commit-sha": "bfa0d53cbe0e9186a17b6919e180d593e047f70f"
  },
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "d9312c8498f76a3b013efcdbaf7bf7aa30117019"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "ebe9d9e4cdaf834b0365a2b53c1cc0681d1d854f"
  }
]
```

File: constants/constants.go [hunk 4]
```diff
@@ -283,12 +285,18 @@ const (
...
+// SSH Certificate tags
 const (
 	// ActionSSHCertificateScopeTagString contains the SSH certificates scope.
 	ActionSSHCertificateScopeTagString = "action=ssh-certificate-scope"
 )
 
+// User Access tags
+const (
+	// ActionUserAccesssInfoTagString contains the info of User Access policy.
+	ActionUserAccesssInfoTagString = "action=user-access-info"
```
Comment (primalmotion) [inline constants/constants.go]: typo.

Also it should be something like `action=restrict-user-access`

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "8dbb7a2094570d2cfd153be7144da6a91967f937"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "272a0f0592f52ab302c22c778c64ae285c31d1a1"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "6073adaaeecf58a3ad30e32b54eaeabe5328e104"
  },
  {
    "project": "sudopolicy",
    "component": "clad",
    "pr-id": "1163",
    "commit-sha": "7dda8f53e5d5d1b8f9635876ec3a48aece11cb2f"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "df4ab1902f3b5a267bb4b8cc466685a0f80aeb55"
  },
  {
    "project": "sudopolicy",
    "component": "domingo",
    "pr-id": "70",
    "commit-sha": "bfa0d53cbe0e9186a17b6919e180d593e047f70f"
  }
]
```

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "8dbb7a2094570d2cfd153be7144da6a91967f937"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "272a0f0592f52ab302c22c778c64ae285c31d1a1"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "6073adaaeecf58a3ad30e32b54eaeabe5328e104"
  },
  {
    "project": "sudopolicy",
    "component": "clad",
    "pr-id": "1163",
    "commit-sha": "7dda8f53e5d5d1b8f9635876ec3a48aece11cb2f"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "b3727af4a2ccaa2e06df1f207e37126dd5b5ede1"
  },
  {
    "project": "sudopolicy",
    "component": "domingo",
    "pr-id": "70",
    "commit-sha": "bfa0d53cbe0e9186a17b6919e180d593e047f70f"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "domingo",
    "pr-id": "70",
    "commit-sha": "bfa0d53cbe0e9186a17b6919e180d593e047f70f"
  },
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "8dbb7a2094570d2cfd153be7144da6a91967f937"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "272a0f0592f52ab302c22c778c64ae285c31d1a1"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "6073adaaeecf58a3ad30e32b54eaeabe5328e104"
  },
  {
    "project": "sudopolicy",
    "component": "clad",
    "pr-id": "1163",
    "commit-sha": "7dda8f53e5d5d1b8f9635876ec3a48aece11cb2f"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "b3727af4a2ccaa2e06df1f207e37126dd5b5ede1"
  }
]
```

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "clad",
    "pr-id": "1163",
    "commit-sha": "7dda8f53e5d5d1b8f9635876ec3a48aece11cb2f"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "5c0331240b2c6ea7b1b06cfdb5b79d1f66851637"
  },
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "8dbb7a2094570d2cfd153be7144da6a91967f937"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "272a0f0592f52ab302c22c778c64ae285c31d1a1"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "6073adaaeecf58a3ad30e32b54eaeabe5328e104"
  }
]
```

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "6073adaaeecf58a3ad30e32b54eaeabe5328e104"
  },
  {
    "project": "sudopolicy",
    "component": "clad",
    "pr-id": "1163",
    "commit-sha": "7dda8f53e5d5d1b8f9635876ec3a48aece11cb2f"
  },
  {
    "project": "sudopolicy",
    "component": "enforcerd",
    "pr-id": "1232",
    "commit-sha": "43f894786e5dec4df2271291b18ee781c105350d"
  },
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "8dbb7a2094570d2cfd153be7144da6a91967f937"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "272a0f0592f52ab302c22c778c64ae285c31d1a1"
  }
]
```

**Comment by t00f** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "sudopolicy",
    "component": "gaia",
    "pr-id": "340",
    "commit-sha": "9fb43cf571626d3551ae2a59035590600ebfbd3a"
  },
  {
    "project": "sudopolicy",
    "component": "backend",
    "pr-id": "428",
    "commit-sha": "090fae12321533aa23a19ffe1b490ad4c40729a6"
  },
  {
    "project": "sudopolicy",
    "component": "underwater",
    "pr-id": "65",
    "commit-sha": "901923ca9e602a92d93a7791494d7b32d5a2b0ba"
  },
  {
    "project": "sudopolicy",
    "component": "clad",
    "pr-id": "1163",
    "commit-sha": "7dda8f53e5d5d1b8f9635876ec3a48aece11cb2f"
  }
]
```

**Pull request merged**